includes/service-providers/constant_contact/class-newspack-newsletters-constant-contact-sdk.php

```diff
@@ -448,4 +459,167 @@ public function create_schedule( $campaign_activity_id, $date = '0' ) {
+	}
+
+	/**
+	 * Create or update a custom field if the type has changed.
+	 *
+	 * @param string $label Custom field label.
+	 * @param string $type  Custom field type. Either 'string' or 'date', defaults
+	 *                      to 'string'. Leave empty to not alter existing type.
+	 *
+	 * @return string Custom field ID.
+	 */
+	public function upsert_custom_field( $label, $type = '' ) {
+		$custom_fields    = $this->get_custom_fields();
+		$custom_field_idx = array_search( $label, array_column( $custom_fields, 'label' ) );
+		if ( false !== $custom_field_idx ) {
+			$custom_field = $custom_fields[ $custom_field_idx ];
+			if ( empty( $type ) || $custom_field->type === $type ) {
+				return $custom_field->custom_field_id;
+			}
+			$this->request(
+				'PUT',
+				'contact_custom_fields/' . $custom_field->custom_field_id,
+				[ 'body' => wp_json_encode( [ 'type' => $type ] ) ]
+			);
+		} else {
+			$custom_field = $this->request(
+				'POST',
+				'contact_custom_fields',
+				[
+					'body' => wp_json_encode(
+						[
+							'label' => $label,
+							'type'  => empty( $type ) ? 'string' : $type,
+						]
+					),
+				]
+			);
+		}
+		return $custom_field->custom_field_id;
+	}
+
+	/**
+	 * Create or update a contact
+	 *
+	 * @param string $email_address Email address.
+	 * @param array  $data          {
+	 *   Contact data.
+	 *
+	 *   @type string   $first_name    First name.
+	 *   @type string   $last_name     Last name.
+	 *   @type string[] $list_ids      List IDs to add the contact to.
+	 *   @type string[] $custom_fields Custom field values keyed by their label.
+	 * }
+	 *
+	 * @return array Created contact data.
+	 */
+	public function upsert_contact( $email_address, $data = [] ) {
+		$contact = $this->get_contact( $email_address );
+		$body    = [];
+		if ( $contact ) {
+			$body = [
+				'email_address'    => get_object_vars( $contact->email_address ),
+				'list_memberships' => $contact->list_memberships,
+				'custom_fields'    => array_map( 'get_object_vars', $contact->custom_fields ),
+				'update_source'    => 'Contact',
+			];
+		} else {
+			$body = [
+				'email_address' => [
+					'address'            => $email_address,
+					'permission_to_send' => 'implicit',
+				],
+				'create_source' => 'Contact',
+			];
+		}
+		if ( ! empty( $data ) ) {
+			if ( isset( $data['first_name'] ) ) {
+				$body['first_name'] = $data['first_name'];
+			}
+			if ( isset( $data['last_name'] ) ) {
+				$body['last_name'] = $data['last_name'];
+			}
+			if ( ! empty( $data['list_ids'] ) ) {
+				if ( ! isset( $body['list_memberships'] ) ) {
+					$body['list_memberships'] = [];
+				}
+				if ( is_string( $data['list_ids'] ) ) {
+					$data['list_ids'] = [ $data['list_ids'] ];
+				}
+				$body['list_memberships'] = array_unique( array_merge( $body['list_memberships'], array_map( 'strval', $data['list_ids'] ) ), SORT_REGULAR );
+			}
+			if ( ! empty( $data['custom_fields'] ) ) {
+				if ( ! $body['custom_fields'] ) {
```

Comment:
Should be:

```suggestion
				if ( ! isset( $body['custom_fields'] ) ) {
```

Otherwise I get `PHP Warning:  Undefined array key "custom_fields"`.